src/scripts/external/mastercrafted-api.ts

```diff
@@ -0,0 +1,235 @@
+    products?: ProductMasterCrafted[];
+    tools?: string[];
+    sound?: string;
+    ownership?: Record<string, unknown>;
+    ingredientsInspection?: 0 | 1;
+    productInspection?: 0 | 1;
+    img?: string;
+}
+
+/**
+ * @typedef {Object} ComponentMasterCrafted
+ * @property {string} [id=undefined]
+ * @property {string} [uuid=undefined]
+ * @property {number} [quantity=undefined]
+ * @property {string} [name=undefined]
+ */
+interface ComponentMasterCrafted {
+    id?: string;
+    uuid?: string;
+    quantity?: number;
+    name?: string;
+}
+
+/**
+ * @typedef {Object} IngredientMasterCrafted
+ * @property {string} id
+ * @property {string} [name=null]
+ * @property {ComponentMasterCrafted[]} [components=[]]
+ * @property {RecipeMasterCrafted} [recipe=null]
+ */
+interface IngredientMasterCrafted {
+    id: string;
+    name?: string;
+    components?: ComponentMasterCrafted[];
+    recipe?: RecipeMasterCrafted;
+}
+
+/**
+ * @typedef {Object} ProductMasterCrafted
+ * @property {string} id
+ * @property {string} name
+ * @property {ComponentMasterCrafted[]} components
+ */
+interface ProductMasterCrafted {
+    id: string;
+    name: string;
+    components: ComponentMasterCrafted[];
+}
+
+
+
+class MasterCraftedAPI extends DefaultRecipeAPI implements RecipeAPI {
+
+    constructor({
+        notificationService,
+        localizationService,
+        recipeValidator,
+        recipeStore,
+        identityFactory,
+        gameProvider,
+        documentManager,
+    }: {
+        notificationService: NotificationService;
+        localizationService: LocalizationService;
+        recipeValidator: RecipeValidator;
+        recipeStore: EntityDataStore<RecipeJson, Recipe>;
+        identityFactory: IdentityFactory;
+        gameProvider: GameProvider;
+        documentManager: DocumentManager;
+    }) {
+        super({
+            notificationService,
+            localizationService,
+            recipeValidator,
+            recipeStore,
+            identityFactory,
+            gameProvider,
+            documentManager
+        })
+    }
+
+    async importRecipeBookFile(recipeBookMasterCraftedPathToJson:string): Promise<Recipe[]> {
+        let recipeBookMasterCrafted: RecipeBookMasterCrafted = JSON.parse(fs.readFileSync(recipeBookMasterCraftedPathToJson, 'utf-8'));
+        return await this.importRecipeBook(recipeBookMasterCrafted);
+    }
+
+    async importRecipeBookJson(recipeBookMasterCraftedJson:string): Promise<Recipe[]> {
+        let recipeBookMasterCrafted: RecipeBookMasterCrafted = JSON.parse(recipeBookMasterCraftedJson);
+        return await this.importRecipeBook(recipeBookMasterCrafted);
+    }
+
+    async importRecipeBook(recipeBookMasterCrafted: RecipeBookMasterCrafted): Promise<Recipe[]> {
+
+        const itemUuids: string[] = recipeBookMasterCrafted.recipes.map((recipe) => {
+            return recipe.id;
+        })
+        const craftingSystemId: string = recipeBookMasterCrafted.id;
```

Comment:
Whilst this does the job, I think I'd like to add some structured data - either in the import/export process, or metadata on the crafting system itself, that explicitly preserves these foreign key relationships with external sources.

Reply:
Mostly I need to understand the desired behaviour for import/export from mastercrafted at this stage, and what the desired cardinality would be between Recipe Books and Crafting systems. 

I'm going to have to look into recipe books and how they are used, but I'm also interested in hearing from you on how you use them. For example, how many do you have, when and why would you create new ones, what do you use them for etc?